trinity/nodes/base.py

```diff
@@ -116,36 +114,43 @@ def notify_resource_available(self) -> None:
             resource_type=BaseChain
         ))
 
-    def make_ipc_server(self) -> Union[IPCServer, EmptyService]:
-        if self._jsonrpc_ipc_path:
+    @property
+    def has_ipc_server(self) -> bool:
+        return bool(self._jsonrpc_ipc_path)
+
+    def make_ipc_server(self, loop: asyncio.AbstractEventLoop) -> BaseService:
+        if self.has_ipc_server:
             rpc = RPCServer(self.get_chain(), self.get_peer_pool())
-            return IPCServer(rpc, self._jsonrpc_ipc_path)
+            return IPCServer(rpc, self._jsonrpc_ipc_path, loop=loop)
         else:
-            return EmptyService()
+            return None
 
     async def _run(self) -> None:
-        self._ipc_server = self.make_ipc_server()
-        # The RPC server needs its own thread, because it provides a synchcronous
-        # API which might call into p2p async methods. These sync->async calls
-        # deadlock if they are run in the same Thread and loop.
-        ipc_loop = self._make_new_loop_thread()
+        if self.has_ipc_server:
+            # The RPC server needs its own thread, because it provides a synchcronous
+            # API which might call into p2p async methods. These sync->async calls
+            # deadlock if they are run in the same Thread and loop.
+            ipc_loop = self._make_new_loop_thread()
 
-        # keep a copy on self, for debugging
-        self._ipc_loop = ipc_loop
+            self._ipc_server = self.make_ipc_server(ipc_loop)
 
-        # FIXME: EmptyService doesn't share a common API with the IPCServer
-        asyncio.run_coroutine_threadsafe(
-            self._ipc_server.run(loop=ipc_loop), loop=ipc_loop  # type: ignore
-        )
+            # keep a copy on self, for later shutdown
+            self._ipc_loop = ipc_loop
+
+            asyncio.run_coroutine_threadsafe(self._ipc_server.run(), loop=ipc_loop)
 
         for service in self._auxiliary_services:
             asyncio.ensure_future(service.run())
 
         await self.get_p2p_server().run()
 
     async def _cleanup(self) -> None:
-        if isinstance(self._ipc_server, IPCServer):
-            await self._ipc_server.stop()
+        # IPC Server requires special handling because it's running in its own loop & thread
+        if self.has_ipc_server:
+            self._ipc_server.threadsafe_cancel()
+            # trigger ipc server thread to close out
```

Comment:
Maybe clarify a bit as it wasn't immediately clear what this meant.

> By stopping the event loop for the ipc server the thread running the server will exit.

Something like that?